ext/compressors/zstd/zstd_compress.c

```diff
@@ -192,15 +296,104 @@ zstd_pre_size(
     return (0);
 }
 
+/*
+ *  zstd_init_context_pool --
+ *	Initialize a given type of context pool.
+ */
+static int
+zstd_init_context_pool(
+  ZSTD_COMPRESSOR *zcompressor, CONTEXT_TYPE ctx_type, int count, ZSTD_CONTEXT_POOL **ctx_pool)
+{
+    WT_EXTENSION_API *wt_api;
+    ZSTD_CONTEXT *context;
+    ZSTD_CONTEXT_POOL *context_pool;
+    int i, ret;
+
+    wt_api = zcompressor->wt_api;
+
+    /* Allocate and initialize both the context pools. */
+    if ((context_pool = calloc(1, sizeof(ZSTD_CONTEXT_POOL))) == NULL)
+        return (errno);
+
+    if ((ret = wt_api->spin_init(wt_api, NULL, &(context_pool->list_lock), "zstd context")) != 0) {
+        (void)wt_api->err_printf(
+          wt_api, NULL, "zstd_init_context_pool: %s", wt_api->strerror(wt_api, NULL, ret));
+        return (ret);
+    }
+    context_pool->count = 0;
+    context_pool->free_ctx_list = NULL;
+
+    for (i = 0; i < count; i++) {
+        context = NULL;
+        if ((context = calloc(1, sizeof(ZSTD_CONTEXT))) == NULL) {
+            (void)wt_api->err_printf(
+              wt_api, NULL, "zstd_init_context_pool: context calloc failure");
+            return (errno);
+        }
+
+        if (ctx_type == CONTEXT_TYPE_COMPRESS)
+            context->ctx = (void *)ZSTD_createCCtx();
+        else
+            context->ctx = (void *)ZSTD_createDCtx();
+
+        if (context->ctx == NULL) {
+            (void)wt_api->err_printf(
+              wt_api, NULL, "zstd_init_context_pool: context create failure");
+            return (errno);
+        }
+        context->next = context_pool->free_ctx_list;
+        context_pool->free_ctx_list = context;
+        context_pool->count++;
+    }
+
+    *ctx_pool = context_pool;
+    return (0);
+}
+
+/*
+ *  zstd_terminate_context_pool --
+ *	Terminate the given context pool.
+ */
+static void
+zstd_terminate_context_pool(
+  WT_COMPRESSOR *compressor, WT_SESSION *session, ZSTD_CONTEXT_POOL **ctx_pool)
+{
+    WT_EXTENSION_API *wt_api;
+    ZSTD_CONTEXT *context;
+    ZSTD_CONTEXT_POOL *context_pool;
+    int i;
+
+    wt_api = ((ZSTD_COMPRESSOR *)compressor)->wt_api;
+    context_pool = *ctx_pool;
+
+    for (i = 0; i < context_pool->count; i++) {
+        context = context_pool->free_ctx_list;
+        context_pool->free_ctx_list = context->next;
+        ZSTD_freeCCtx((ZSTD_CCtx *)context->ctx);
```

Comment:
I thought I'd see calls to ZSTD_freeDCtx somewhere here - does the API make it so that's not necessary?